st2actions/st2actions/policies/retry.py

```diff
@@ -0,0 +1,141 @@
+        :param retry_on: Condition to retry the execution on.
+        :type retry_on: ``str``
+
+        :param max_retry_count: Maximum number of times to try to retry an action.
+        :type max_retry_count: ``int``
+        """
+        # TODO: Also support sleep delay before retrying
+        super(ExecutionRetryPolicyApplicator, self).__init__(policy_ref=policy_ref,
+                                                             policy_type=policy_type)
+
+        if retry_on not in RetryOnPolicy.get_valid_values():
+            raise ValueError('Invalid value for "retry_on" parameter')
+
+        if max_retry_count > MAXIMUM_RETRY_COUNT:
+            raise ValueError('Maximum retry count is %s' % (MAXIMUM_RETRY_COUNT))
+
+        self.retry_on = retry_on
+        self.max_retry_count = max_retry_count
+
+    def apply_before(self, target):
+        # Nothing to do here
+        target = super(ExecutionRetryPolicyApplicator, self).apply_before(target=target)
+        return target
+
+    def apply_after(self, target):
+        target = super(ExecutionRetryPolicyApplicator, self).apply_before(target=target)
+
+        live_action_db = target
+        retry_count = self._get_live_action_retry_count(live_action_db=live_action_db)
+
+        extra = {'live_action_db': live_action_db, 'policy_ref': self._policy_ref,
+                 'retry_on': self.retry_on, 'max_retry_count': self.max_retry_count,
+                 'current_retry_count': retry_count}
+
+        if live_action_db.status not in VALID_RETRY_STATUSES:
+            # Currently we only support retrying on failed action
+            LOG.debug('Liveaction not in a valid retry state, not checking retry policy',
+                      extra=extra)
+            return target
+
+        if (retry_count + 1) > self.max_retry_count:
+            LOG.info('Maximum retry count has been reached, not retrying', extra=extra)
+            return target
+
+        has_failed = live_action_db.status == LIVEACTION_STATUS_FAILED
+        has_timed_out = live_action_db.status == LIVEACTION_STATUS_TIMED_OUT
+
+        if has_failed and self.retry_on == RetryOnPolicy.FAILURE:
+            extra['failure'] = True
+            LOG.info('Policy matched (failure), retrying action execution...', extra=extra)
+            self._re_run_live_action(live_action_db=live_action_db)
+            return target
+
+        if has_timed_out and self.retry_on == RetryOnPolicy.TIMEOUT:
+            extra['timeout'] = True
+            LOG.info('Policy matched (timeout), retrying action execution...', extra=extra)
+            self._re_run_live_action(live_action_db=live_action_db)
+            return target
+
+        return target
+
+    def _get_live_action_retry_count(self, live_action_db):
+        """
+        Retrieve current retry count for the provided live action.
+
+        :rtype: ``int``
+        """
+        # TODO: Ideally we would store retry_count in zookeeper or similar and use locking so we
```

Comment:
My initial reaction here is that the retry_context does not belong to a single execution. Keeping it separate does have the extra benefits of being able to manage this information better. However, clients that care to demonstrate this information will then have to make multiple requests which can be bandaged over by using the views endpoint.